File: examples/go/geocode.go
// CSV2GEO API - Go Geocoding Example
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
)

const (
	apiKey  = "YOUR_API_KEY"
	baseURL = "https://api.csv2geo.com/v1"
)

// Location represents a geographic coordinate
type Location struct {
	Lat float64 `json:"lat"`
	Lng float64 `json:"lng"`
}

// AddressComponents contains parsed address parts
type AddressComponents struct {
	HouseNumber string `json:"house_number"`
	Street      string `json:"street"`
	City        string `json:"city"`
	State       string `json:"state"`
	Postcode    string `json:"postcode"`
	Country     string `json:"country"`
}

// GeocodeResult represents a single geocoding result
type GeocodeResult struct {
	FormattedAddress string            `json:"formatted_address"`
	Location         Location          `json:"location"`
	Accuracy         string            `json:"accuracy"`
	Components       AddressComponents `json:"components"`
}

// GeocodeResponse is the API response for geocoding
type GeocodeResponse struct {
	Query   string          `json:"query"`
	Results []GeocodeResult `json:"results"`
}

// Geocode converts an address to coordinates
func Geocode(address string) (*GeocodeResponse, error) {
	params := url.Values{}
	params.Add("q", address)
	params.Add("api_key", apiKey)

	resp, err := http.Get(fmt.Sprintf("%s/geocode?%s", baseURL, params.Encode()))
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("geocoding failed: %s", string(body))
	}

	var result GeocodeResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &result, nil
}

// ReverseGeocode converts coordinates to an address
func ReverseGeocode(lat, lng float64) (*GeocodeResponse, error) {
	params := url.Values{}
	params.Add("lat", fmt.Sprintf("%f", lat))
	params.Add("lng", fmt.Sprintf("%f", lng))
	params.Add("api_key", apiKey)

	resp, err := http.Get(fmt.Sprintf("%s/reverse?%s", baseURL, params.Encode()))
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("reverse geocoding failed: %s", string(body))
	}

	var result GeocodeResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &result, nil
}

// BatchGeocodeRequest is the request body for batch geocoding
type BatchGeocodeRequest struct {
	Addresses []string `json:"addresses"`
}

// BatchGeocodeResponse is the response for batch geocoding
type BatchGeocodeResponse struct {
	Results []GeocodeResponse `json:"results"`
}

// BatchGeocode geocodes multiple addresses
func BatchGeocode(addresses []string) (*BatchGeocodeResponse, error) {
	reqBody, err := json.Marshal(BatchGeocodeRequest{Addresses: addresses})
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequest("POST", baseURL+"/geocode", bytes.NewBuffer(reqBody))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+apiKey)

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("batch geocoding failed: %s", string(body))
	}

	var result BatchGeocodeResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &result, nil
}

func main() {
	// Forward geocoding
	fmt.Println("Forward Geocoding:")
	result, err := Geocode("1600 Pennsylvania Ave, Washington DC")
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}
	fmt.Printf("  Address: %s\n", result.Results[0].FormattedAddress)
	fmt.Printf("  Lat: %f\n", result.Results[0].Location.Lat)
	fmt.Printf("  Lng: %f\n", result.Results[0].Location.Lng)

	// Reverse geocoding
	fmt.Println("\nReverse Geocoding:")
	reverseResult, err := ReverseGeocode(38.8977, -77.0365)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}
	fmt.Printf("  Address: %s\n", reverseResult.Results[0].FormattedAddress)

	// Batch geocoding
	addresses := []string{
		"1600 Pennsylvania Ave, Washington DC",
		"350 Fifth Avenue, New York, NY",
		"1 Infinite Loop, Cupertino, CA",
	}
	batchResult, err := BatchGeocode(addresses)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}
	fmt.Printf("\nBatch Geocoding: %d addresses processed\n", len(batchResult.Results))
}

File: examples/php/geocode.php
<?php
/**
 * CSV2GEO API - PHP Geocoding Example
 */

define('API_KEY', 'YOUR_API_KEY');
define('BASE_URL', 'https://api.csv2geo.com/v1');

/**
 * Forward geocode an address to coordinates
 *
 * @param string $address The address to geocode
 * @return array Geocoding result
 */
function geocode(string $address): array
{
    $params = http_build_query([
        'q' => $address,
        'api_key' => API_KEY
    ]);

    $response = file_get_contents(BASE_URL . "/geocode?{$params}");

    if ($response === false) {
        throw new Exception('Geocoding request failed');
    }

    return json_decode($response, true);
}

/**
 * Reverse geocode coordinates to an address
 *
 * @param float $lat Latitude
 * @param float $lng Longitude
 * @return array Address result
 */
function reverseGeocode(float $lat, float $lng): array
{
    $params = http_build_query([
        'lat' => $lat,
        'lng' => $lng,
        'api_key' => API_KEY
    ]);

    $response = file_get_contents(BASE_URL . "/reverse?{$params}");

    if ($response === false) {
        throw new Exception('Reverse geocoding request failed');
    }

    return json_decode($response, true);
}

/**
 * Batch geocode multiple addresses
 *
 * @param array $addresses Array of addresses (max 10,000)
 * @return array Batch results
 */
function batchGeocode(array $addresses): array
{
    $ch = curl_init(BASE_URL . '/geocode');

    curl_setopt_array($ch, [
        CURLOPT_POST => true,
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_HTTPHEADER => [
            'Content-Type: application/json',
            'Authorization: Bearer ' . API_KEY
        ],
        CURLOPT_POSTFIELDS => json_encode(['addresses' => $addresses])
    ]);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    curl_close($ch);

    if ($httpCode !== 200) {
        throw new Exception("Batch geocoding failed with status {$httpCode}");
    }

    return json_decode($response, true);
}

// Example usage
try {
    // Forward geocoding
    echo "Forward Geocoding:\n";
    $result = geocode('1600 Pennsylvania Ave, Washington DC');
    echo "  Address: {$result['results'][0]['formatted_address']}\n";
    echo "  Lat: {$result['results'][0]['location']['lat']}\n";
    echo "  Lng: {$result['results'][0]['location']['lng']}\n";

    // Reverse geocoding
    echo "\nReverse Geocoding:\n";
    $reverseResult = reverseGeocode(38.8977, -77.0365);
    echo "  Address: {$reverseResult['results'][0]['formatted_address']}\n";

    // Batch geocoding
    $addresses = [
        '1600 Pennsylvania Ave, Washington DC',
        '350 Fifth Avenue, New York, NY',
        '1 Infinite Loop, Cupertino, CA'
    ];
    $batchResult = batchGeocode($addresses);
    echo "\nBatch Geocoding: " . count($batchResult['results']) . " addresses processed\n";

} catch (Exception $e) {
    echo "Error: " . $e->getMessage() . "\n";
}
